client/src/components/Map/Map.js

```diff
@@ -18,10 +31,57 @@ const propTypes = {
       })
     })
   ),
-  nodes: PropTypes.arrayOf(ChallengeNode)
+  isInitialized: PropTypes.bool,
+  nodes: PropTypes.arrayOf(ChallengeNode),
+  toggleBlock: PropTypes.func.isRequired,
+  toggleSuperBlock: PropTypes.func.isRequired
 };
 
-class ShowMap extends Component {
+const mapStateToProps = state => {
+  return createSelector(
+    currentChallengeIdSelector,
+    currentChallengeUrlSelector,
+    isInitializedSelector,
+    (currentChallengeId, currentChallengeUrl, isInitialized) => ({
+      currentChallengeId,
+      currentChallengeUrl,
+      isInitialized
+    })
+  )(state);
+};
+
+function mapDispatchToProps(dispatch) {
+  return bindActionCreators(
+    {
+      toggleSuperBlock,
+      toggleBlock
+    },
+    dispatch
+  );
+}
+
+export class Map extends Component {
+  constructor(props) {
+    super(props);
+    // Tries to use the cookie, then the store value and finally defaults
+    // to the first challenge.
+    const currentChallengeUrl =
+      cookies.get('currentChallengeUrl') ||
+      props.currentChallengeUrl ||
+      props.nodes[0].fields.slug;
+
+    if (!this.props.isInitialized)
```

Comment:
I still don't see why should we persist expanded state during the session?

Reply:
It's a bit of an edge case, but that is a fallback if the user disabled cookies (or local storage since we're going to use that).  It won't matter often.

Reply:
I went through lessons, then return to the Map, and I'll get state like I have not done anything.

Reply:
Yeah, that's because I wasn't sure if it was better to expand the Map to the current challenge each time the user goes there or only when they reload the page.  It's easy enough to change if that's the desired behaviour.

Reply:
@raisedadead besides this, LGTM

Reply:
I would prefer if we expand each time user went there.

Reply:
What do you mean, expand the current challenge or last state?

Reply:
@ValeraS @raisedadead This latest commit should do the trick.

Reply:
It's the trick the right solution will be using local state.

Reply:
> I would prefer if we expand each time user went there.

I would prefer to expand to whatever last challenge I was on before clicking the `learn` (Map) on the nav, with a fallback to whatever we have in the state locally.

Reply:
I understood @raisedadead's comment
> I would prefer if we expand each time user went there.

to mean he wanted the `Map` to expand to show the last viewed challenge whenever you visited it, rather than just after reloading the page.  Which is now what happens.

> It's the trick right solution will be using local state.

It *does* use local state; it uses `currentChallengeId`.

Reply:
It uses the store for currently open sections but should use local state.

Reply:
The blocks and superblocks need to be able to be able to manipulate the state, which is rather easy if they interact with the store.  What's wrong with doing it that way?

Also, the point of this PR was to add a feature.  If we want to further refactor things, why not do that in a separate PR?

Reply:
I only said it's a trick because now you need to reset state in the store on each creation. By the logic, it's a local state not global. if you want, we can fix it in another PR.

Reply:
Ah, I see what you mean.  I think 'does the trick' was a poor choice of phrase by me.  It's an idiom meaning "achieves the goal".  Anyways, I see why resetting it is kind of a trick.  Perhaps we can come up with a nice, local version, later.